compiler/src/dotty/tools/dotc/transform/ElimRepeated.scala

```diff
@@ -54,11 +55,22 @@ class ElimRepeated extends MiniPhase with InfoTransformer { thisPhase =>
   private def elimRepeated(tp: Type)(using Context): Type = tp.stripTypeVar match
     case tp @ MethodTpe(paramNames, paramTypes, resultType) =>
       val resultType1 = elimRepeated(resultType)
-      val paramTypes1 =
-        if paramTypes.nonEmpty && paramTypes.last.isRepeatedParam then
-          val last = paramTypes.last.translateFromRepeated(toArray = tp.isJavaMethod)
-          paramTypes.init :+ last
-        else paramTypes
+      val paramTypes1 = paramTypes match
+        case init :+ last if last.isRepeatedParam =>
+          val isJava = tp.isJavaMethod
+          // A generic Java varargs `T...` is erased to `Object[]` in bytecode,
+          // we directly translate such a type to `Array[_ <: Object]` instead
+          // of `Array[_ <: T]` here. This allows the tree transformer of this phase
+          // to emit the correct adaptation for repeated arguments without
+          // relying on any cast (cf `adaptToArray`).
+          val last1 =
+            if isJava && last.elemType.isInstanceOf[TypeParamRef] then
+              defn.ArrayOf(TypeBounds.upper(defn.ObjectType))
```

Comment:
If `T` has a bound that is smaller than `Object`, for example `<T extends java.util.List>`, shouldn't this be translated to `Array[_ <: List]` instead of `Array[_ <: Object]`?

As a concrete example: `java.util.Arrays.parallelSort(T[])` has the following signature and descriptor:
```java
  public static <T extends java.lang.Comparable<? super T>> void parallelSort(T[]);
    descriptor: ([Ljava/lang/Comparable;)V
```
(shown by `javap -s java.util.Arrays`)

Reply:
Right, I simplified that condition too much, we should only to do this translation when T is unbounded (that's the only situation in which we can end up with an array of primitive that will need to be adapted).